src/AspNetCoreModuleV2/CommonLib/StdWrapper.cpp

```diff
@@ -0,0 +1,124 @@
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+#include "stdafx.h"
+#include "StdWrapper.h"
+#include "exceptions.h"
+#include "LoggingHelpers.h"
+#include <corecrt_io.h>
+#include <fcntl.h>
+
+StdWrapper::StdWrapper(FILE* stdStream, DWORD stdHandleNumber, HANDLE handleToRedirectTo, BOOL fEnableNativeRedirection)
+    : m_previousFileDescriptor(0),
+        m_stdStream(stdStream),
+        m_stdHandleNumber(stdHandleNumber),
+        m_fEnableNativeRedirection(fEnableNativeRedirection),
+        m_handleToRedirectTo(handleToRedirectTo),
+        m_redirectedFile(nullptr)
+{
+}
+
+StdWrapper::~StdWrapper() = default;
+
+// Redirects stdout/stderr to the provided handle.
+// Example:
+// If the handleToRedirecTo = 0x24
+// Before:
+// _fileno(stdout) = 1
+// GetStdHandle(STD_OUTPUT_HANDLE) = 0x20
+// After:
+// _fileno(stdout) = 3
+// GetStdHandle(STD_OUTPUT_HANDLE) = 0x28 <- Duplicated from 0x24
+HRESULT 
+StdWrapper::StartRedirection()
+{
+    HANDLE stdHandle;
+
+    // In IIS, stdout and stderr are set to null as w3wp is created with DETACHED_PROCESS,
+    // so fileno(m_stdStream) returns -2.
+    // Open a null file such that undoing the redirection succeeds and _dup2 works.
+    // m_previousFileDescriptor will be used for restoring stdout/stderr
+    if (_fileno(m_stdStream) == -2)
+    {
+        freopen_s((FILE**)&m_stdStream, "nul", "w", m_stdStream);
+        m_previousFileDescriptor = _dup(_fileno(m_stdStream));
+    }
+    else
+    {
+        m_previousFileDescriptor = _dup(_fileno(m_stdStream));
+    }
+
+    // After setting the std handle, we need to set stdout/stderr to the current
+    // output/error handle.
+
+    // Duplicate the handle before opening the handle and associating a file pointer.
+    // If we don't by calling close on the file, the same handle value will be closed
+    // multiple times.
+    // Note, by calling duplicate handle, the new handle returned will have a different value
+    // than the original, but point to the same underlying file object.
+    RETURN_LAST_ERROR_IF(!DuplicateHandle(
+        /* hSourceProcessHandle*/ GetCurrentProcess(),
+        /* hSourceHandle */ m_handleToRedirectTo,
+        /* hTargetProcessHandle */ GetCurrentProcess(),
+        /* lpTargetHandle */&stdHandle,
+        /* dwDesiredAccess */ 0, // dwDesired is ignored if DUPLICATE_SAME_ACCESS is specified
+        /* bInheritHandle */ TRUE,
+        /* dwOptions  */ DUPLICATE_SAME_ACCESS));
+
+    SetStdHandle(m_stdHandleNumber, stdHandle);
+
+    if (!m_fEnableNativeRedirection)
+    {
+        return S_OK;
+    }
+
+    // _open_osfhandle will associate a filedescriptor with the handle.
+    const auto fileDescriptor = _open_osfhandle(reinterpret_cast<intptr_t>(stdHandle), _O_WTEXT | _O_TEXT);
+
+    RETURN_LAST_ERROR_IF(fileDescriptor == -1);
+
+    m_redirectedFile = _fdopen(fileDescriptor, "w");
+
+    RETURN_LAST_ERROR_IF(m_redirectedFile == nullptr);
```

Comment:
Does _fdopen set lasterror? I thought it sets errno.

Reply:
I'll go through and double check these. I think you are right.